staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -1929,6 +1929,95 @@ func TestCELValidationContextCancellation(t *testing.T) {
 	}
 }
 
+func getRecursionRule(exp string, recDep int) string {
+	if recDep <= 0 {
+		return exp
+	}
+	res := exp
+	for i := 1; i < recDep; i++ {
+		res += "==" + exp
+	}
+	return res
+}
+
+// TestCELMaxRecursionDepth tests CEL setting for maxRecursionDepth.
+func TestCELMaxRecursionDepth(t *testing.T) {
+	tests := []struct {
+		name          string
+		schema        *schema.Structural
+		obj           interface{}
+		oldObj        interface{}
+		valid         []string
+		errors        map[string]string // rule -> string that error message must contain
+		costBudget    int64
+		isRoot        bool
+		expectSkipped bool
+	}{
+		{name: "test CEL maxRecursionDepth",
+			obj:    objs(true),
+			schema: schemas(booleanType),
+			valid: []string{
+				getRecursionRule("self.val1", 243),
```

Comment:
huh... I expected 250 to be valid, and 251 to be invalid... where did the 243 / 244 limit come from?

Reply:
That is also my confusion. I raise it in above comment thread and will work with Tristan on it.

Reply:
The recursion limit is based on the internal recursion in the parser, so the expression recursion at a grammar level will be close to, but not exactly the same as the parser's internal behavior.

Reply:
> The recursion limit is based on the internal recursion in the parser, so the expression recursion at a grammar level will be close to, but not exactly the same as the parser's internal behavior.

oh, I see.

> The important thing is that the limit cannot get smaller, but it could get larger with more improvements in the grammar or ANTLR stack

That's good to know. Is that tested/guaranteed in a test upstream (either in cel or in antlr)?

Reply:
Thanks for the reply! In this case I am wondering if we should monitor it strictly in k8s?  Or we could just check it in lower side and leave rooms for further possible improvement?

Reply:
I do think we want a test around the edges of what is accepted by our API (the largest accepted value and the smallest rejected value). For example, we have similar tests around rejected recursive parsing of JSON and YAML.